### ISSUE TITLE
NaN in loss for MNIST

### ISSUE THREAD

**Issue by luis-armando-perez-rey** [nicola-decao/s-vae-tf#3]
Dear Nicola,
I was trying to train the S-VAE with the binarized MNIST as in the provided example to reproduce the results presented in the paper using:

print()
print('##### Hyper-spherical VAE #####')
for i in range(1000):
    # training
    x_mb, _ = mnist.train.next_batch(64)
    # dynamic binarization
    x_mb = (x_mb > np.random.random(size=x_mb.shape)).astype(np.float32)
    
    session.run(optimizerS.train_step, {modelS.x: x_mb})

    # every 100 iteration plot validation
    if i % 100 == 0:
        x_mb = mnist.validation.images
        # dynamic binarization
        x_mb = (x_mb > np.random.random(size=x_mb.shape)).astype(np.float32)
        
        print(i, session.run({**optimizerS.print}, {modelS.x: x_mb}))

After repeating this command several times (to account for several epochs), the loss function becomes NaN for the KL divergence term. For example:

##### Hyper-spherical VAE #####
0 {'recon loss': 142.977, 'ELBO': -148.5604, 'KL': 5.583393}
100 {'recon loss': 143.57727, 'ELBO': -149.2204, 'KL': 5.643135}
200 {'recon loss': 206.4642, 'ELBO': nan, 'KL': nan}
300 {'recon loss': 206.10231, 'ELBO': nan, 'KL': nan}
400 {'recon loss': 206.28786, 'ELBO': nan, 'KL': nan}
500 {'recon loss': 206.16287, 'ELBO': nan, 'KL': nan}
600 {'recon loss': 206.15532, 'ELBO': nan, 'KL': nan}
700 {'recon loss': 206.27534, 'ELBO': nan, 'KL': nan}
800 {'recon loss': 206.04416, 'ELBO': nan, 'KL': nan}
900 {'recon loss': 206.07002, 'ELBO': nan, 'KL': nan}

This happens always if I run for sufficiently many epochs the S-VAE training. Upon inspection, what I have noticed is that the encoder becomes degenerate and sends any input into a fixed point within the sphere. Is there a way to increase the stability of training? Could we obtain the code used to reproduce the results presented in the paper?

Thank you for your help. 

Best regards,

Luis

**Comment by nicola-decao** [comment]
Does this happen to you also when `DIM` is larger than 2?

**Comment by luis-armando-perez-rey** [comment]
I tried with DIM 3 for for several epochs. I don't get NaN values anymore, but the training stops after several epochs. 
To see what was the problem was I gave to the VonMisesFisher the allow_nan_stats=False which raised an error while training after some epochs. My guess is that the network starts predicting unstable values for kappa. Have you encountered this issue? I am guessing that restricting the values of kappa to a certain range could potentially avoid the issue but I am not sure if this error could be avoided another way.

**Comment by luis-armando-perez-rey** [comment]
While reading other papers I found the following comment about the implementation: 

"While the code used to implement S-VAE worked as intended on MNIST,
on the link prediction task we encountered numerical issues that made it impossible to replicate the results of [1] in our different experimental setting
with different data splits, re-weighting technique for the loss, and hyperparameters searched. When further investigating the instability of S-VAE,
we observed a computational issue in the model, which would saturate the
floating point representation of the GPU, resulting in invalid gradients being
propagated through the network. The numerical instability derives from the
sampling procedure of the von Mises-Fisher distribution on which the model
is based, as the exponentially scaled modified Bessel function used for sampling causes a division by zero for higher values of d." - Adversarial Autoencoders with Constant-Curvature
Latent Manifolds https://arxiv.org/pdf/1812.04314.pdf

I was wondering if you had explored this issue as well and if such explanation given is accurate in this case.
